### DISCUSSION TITLE
Reinventing the DLS API

### DISCUSSION THREAD

**Discussion by gasche** [ocaml/ocaml#13356] [Ideas]
I am unsatisfied with the DLS APIs that we currently have, and also with the one proposed in #12724 and #13355. Opening this discussion to gather ideas and try to converge to a satisfying state.

Current API (in trunk):

```ocaml
module DLS : sig
(** Domain-local Storage *)

    type 'a key
    (** Type of a DLS key *)

    val new_key : ?split_from_parent:('a -> 'a) -> (unit -> 'a) -> 'a key
    (** [new_key f] returns a new key bound to initialiser [f] for accessing
,        domain-local variables. [...] *)

    val get : 'a key -> 'a
    (** [get k] returns [v] if a value [v] is associated to the key [k] on
        the calling domain's domain-local state. Sets [k]'s value with its
        initialiser and returns it otherwise. *)

    val set : 'a key -> 'a -> unit
    (** [set k v] updates the calling domain's domain-local state to associate
        the key [k] with value [v]. It overwrites any previous values associated
        to [k], which cannot be restored later. *)

end
```

Our current replacement candidate, which evolved thanks to the contributions of @dbuenzli, @c-cube and @kayceesrk, looks fairly similar (it is for TLS, not DLS, just replace "thread" by "domain" in your mind):

```ocaml
module Thread_local_storage : sig
  type 'a t
  (** Type of a thread-local storage key *)

  val make: ?split_from_parent:('a -> 'a) -> (unit -> 'a) -> 'a t
  (** [make f] returns a new key bound to initialiser [f] for accessing
      thread-local variables. [...] *)

  val get : 'a t -> 'a
  (** [get k] returns [v] if a value [v] is associated to the key [k] on
      the calling thread's thread-local state. Sets [k]'s value with its
      initialiser and returns it otherwise. *)

  val set : 'a t -> 'a -> unit
  (** [set k v] updates the calling thread's thread-local state to associate
    the key [k] with value [v]. It overwrites any previous values associated
    to [k], which cannot be restored later. *)
end
```

Comments:

- The notion of "tls key" comes from pthread, it is familiar to domain experts. Broadly the idea is that each domain/thread has a single, dynamically-resized table, and the "keys" are indices into this table.
- This API style feels alien in OCaml programs, because it corresponds to indices inside a global, mutable, singleton data structure that is hidden in the API, generally a design antipattern. (We are familiar with `get table key`, but not with `get key`.)
- Dually one can see these "keys" as mutable variables/references/locations themselves, that take a different value for each domain/thread. So instead of a domain-local "storage key", we should call them domain-local "references". (Note how the documentation for `new_key / make` mentions that it creates a domain-local "variable", going in this direction. I think in OCaml we generally try to avoid "variable" for shared mutable state.)
- The candidate API sits incomfortably in the middle, because the `'a t, make, get, set` API clearly suggests the "domain-local reference" view, but the documentation comments mention keys instead.

In addition, there are many small glitches with those APIs. For example the type `'a Thread_local_storage.t` reads badly (unlike `'a Thread_local_storage.key` which makes sense).

My current thinking is that we should try to consistently use one view in the API and its documentation, or maybe to pick one, but document both (describe the few operations under both points of view).

**Comment by dbuenzli** [comment]
> In addition, there are many small glitches with those APIs. For example the type `'a Thread_local_storage.t` reads badly (unlike `'a Thread_local_storage.key` which makes sense).

I can't find the reference again but I think that at some point I suggested we should add the `Ref` module and then have: 

```ocaml
module Ref : sig
  type 'a t = 'a ref
  val make : 'a -> 'a t
  val get : 'a t -> 'a
  val set : 'a t -> 'a -> unit

  module Thread_local : sig 
     type 'a t
     val make : ?from_parent:('a -> 'a) -> (unit -> 'a) -> 'a t
     val get : 'a t -> 'a 
     val set : 'a t -> 'a -> unit
  end
end
```

I think people didn't like it but I don't remember exactly why.

**Reply by gasche** [reply]
I agree that this reads well. Maybe one argument against this presentation is that it "pollutes" a common module with a very specialized submodule for an entirely different problem domain and use-cases.

I was thinking of `Thread_local_ref` and `Domain_local_ref` toplevel stdlib modules, which is fairly close.

```ocaml
module Thread_local_ref : sig 
  type 'a t
  val make : ?from_parent:('a -> 'a) -> (unit -> 'a) -> 'a t
  val get : 'a t -> 'a 
  val set : 'a t -> 'a -> unit
end
```

**Reply by dbuenzli** [reply]
For me it's just mutable state in different computing scopes so I don't really see it as a different "problem" or use-case. 

I see it as a modality on mutable state: `Ref` is a reference shared by every computing scope, `Ref.Thread_local` is a private reference for every thread, `Ref.Domain_local` is a private reference for every domain[^1]. I think the nesting nicely highlights it while avoiding to add too much names to the toplevel scope.

But if those have to be added to the toplevel scope using `_ref` instead of `_storage` is definitively a better fit with the current OCaml terminology. 

I think this "storage" and "key" terminology should just be dropped: if you try to explain to users in the `.mli` preamble what these structure really are (rather than rely on alien jargon as is done now) you would say something like "this is a reference cell, and each thread has its own private copy and can only operate on its own" (can be improved). So perhaps we should start by writing a good explanation of what these structures are for OCaml programmers and the mental model they should have about them. The names will naturally follow.

[^1]: While we are at brainstorming names `Ref.{Domain,Thread}_private` :-) ?

**Reply by c-cube** [reply]
Conceptually I can see the similarity, but in usage and implementation they differ so wildly that I don't like it :-). A `_ ref` is a very lightweight object that might live in as short as a single scope or loop, and might even become only stack or register allocated. TLS/DLS is heavy storage tied to the lifetime of a thread or domain and it must only ever be allocated in advance at the toplevel. So while I think it's ok to use `_ref` for all of them, I don't think it makes sense to have DLS/TLS be submodules of `Ref`.

**Comment by gasche** [comment]
> For me it's just mutable state in different computing scopes so I don't really see it as a different "problem" or use-case.

One example of difference in usage is that OCaml references are meant to be created with abandon, you can easily create thousands of references in your OCaml programs, use them on each node of a large graph, etc. On the other hand each library should define a constant (as in O(1)) number of DLS/TLS keys, otherwise things start blowing up because the implementation optimizes access speed at the expense of memory density. If I gave you a type of mutable cell, but you can only create 2 of them per opam package, you wouldn't call it `Ref`.

**Reply by dbuenzli** [reply]
Right. I would perhaps call it a `global_ref`.

**Reply by dbuenzli** [reply]
So maybe resuing `ref` is not a good idea overall then. It seems you want to find something to distinguish these structures from `ref`s. 

Using `global` is of course going to be confusing but their use is a form of global/toplevel references, except they are private to the computing scope. So you can see them as being additions to the state needed to manage the computing scope. 

So perhaps you could try to use "state" somewhere:

```ocaml
Thread_state.make
Thread_custom_state.make
Custom_thread_state.make
Thread_local_state.make
```
and describe it as "custom mutable state associated to a thread".

**Reply by gasche** [reply]
I also wondered about `Per_{thread,domain}_global`. (But: people familiar with the "thread-local variable" terminology could be reasonably surprised to hear them describes as "per-thread globals".)

**Reply by gasche** [reply]
Instead of `reference` we could use a word which is a synonym, but is used less often, so there are fewer risks of carrying unwanted connotations -- and we can add new ones. I was thinking of "cell", but I also like "slot", which has the mouth-watering quality of preserving the DLS/TLS acronyms.

**Comment by c-cube** [comment]
Thank you for resurrecting this, @gasche. I'm looking forward to the `thread-local-storage` package to only be a stub for a safe, standard mechanism!!

**Comment by gasche** [comment]
My current favorite proposal, based on the discussion here:

```ocaml
module Thread_local_slot : sig 
  type 'a t
  val make : ?from_parent:('a -> 'a) -> (unit -> 'a) -> 'a t
  val get : 'a t -> 'a 
  val set : 'a t -> 'a -> unit
end
```

**Reply by gasche** [reply]
With some documentation comments. (The "Note" is from @kayceesrk, it is part of #12724.)

```ocaml
module Thread_local_slot : sig
  type 'a t
  (** A thread-local slot of type ['a t] is a global mutable data
      structure that may contain a different value on each
      thread. Each thread can only access its own value for the slot,
      so that no synchronization is necessary. The typical use-case
      for thread-local storage is to adapt algorithms and libraries
      relying on global mutable state to work in multi-threaded
      scenarios.

      On a given thread, an ['a slot] may either hold a value of type
      ['a] or be empty. It is paired with an initializer of type
      [unit -> 'a] which may be used to compute a value on first
      access -- see {!get}.

      {b Note:} Thread-local storage is designed for fast,
      constant-time access to a small number of slots, typically
      declared at the top-level of the module. It is not designed to
      create a large numbers of slots dynamically, and it may be slow
      and space-inefficient in that situation.
  *)

  val make: ?split_from_parent:('a -> 'a) -> (unit -> 'a) -> 'a t
  (** [make f] returns a new slot with the initialiser [f];
      it is empty on all threads. *)

  val get : 'a t -> 'a
  (** [get s] returns the value of the slot [s] on the calling thread. If
      the slot does not have a value yet, its initializer is called on
      the thread, and the resulting value is stored in the slot and
      returned. *)

  val set : 'a t -> 'a -> unit
  (** [set s v] sets the value of the slot [s] on the calling thread. It
      overwrites any previous value associated to [s]. *)
end
```

**Reply by dbuenzli** [reply]
A few comments:

1. I remain [adamant](https://github.com/ocaml/ocaml/pull/12724#discussion_r1410507184) that the `split_` is going to read weird for many use cases. 

2. `make`, I guess quite a few things needs to be said here about which thread runs what. I'm not convinced about the interface which I find difficult to read/remember when `from_parent` is used (i.e. IIUC `f` is used by the main thread of a domain and children only ever execute `from_parent`). I think it would be better to either have two `make` functions (like @lpw25 suggested) with properly labelelled functions or a more complex `init` type and a single `make`:
    ```ocaml
    type init = 
    | Inherit of { main : unit -> 'a; child : parent:'a -> 'a } 
    | Init of unit -> a
    ```

4. I'm not entirely sold on `slot` yet (vs `state`) but I could be; at least it's much better than `storage` or `ref`. But in your description I find it a bit weird to call this a "data structure". I don't think I would call a reference cell a "data structure". I would rather describe this as `a value of type ['a t] is global mutable state that…` at which point using `state` (i.e. `Thread_local_state`) is more enticing to me.

**Reply by gasche** [reply]
I don't like the "state" terminology because I tend to think of "state" as a global thing, the state of the whole computation, the state of memory, etc. I would not call a reference cell a "state". When we say that something has "two states", we don't mean two different cells/slots/etc., we mean that the state can take two different values (for yet another slightly different meaning of the word "state", which describes the whole configuration of computation and not just some information on the side that computation accesses.)

**Comment by lpw25** [comment]
This has probably already been discussed somewhere, but are we sure that the interface should provide `set`? It seems just as expressive without since you can use a `ref` as the contents. I think I'd probably be surprised if the indirection mattered for performance in practice. You could potentially get some benefit in terms of CSE as well.

**Reply by dbuenzli** [reply]
I remembered that comment of yours and thought about this a bit over the weekend. 

The only thing I found a little bit less nice is for the case of parent inheritance: if your slot is a `ref` you inherit an `'a ref` from the parent (rather than a `'a`) to transform and thus you can meddle with the parent state at that point. But since it's all tightly defined during `make` you could argue that it's ok since the user of the slot knows what's it is doing.

I think that not having `set` would bring the design even more towards the idea that this is a "thread local toplevel definition" since in OCaml let bindings are immutable by default.

TL;DR I like the idea.

**Reply by c-cube** [reply]
An indirection is never free, and this kind of thing might be called in critical paths (e.g. to find a thread-local logging/tracing facility, or the current fiber, or the current scheduler, etc.). Personally I'd also advocate for toplevel bindings to be (opt in) mutable, except that they're not thread safe, unlike TLS. Why not stick to the same get/set interface as all the other languages, especially if it's also the fast option?

**Reply by lpw25** [reply]
There is only an extra indirection if the value you are storing is not already mutable. The lack of mutation also allows you to safely hold on to the value over calls to user code, which may be a noticeable performance win for some use cases.

> might be called in critical paths (e.g. to find a thread-local logging/tracing facility, or the current fiber, or the current scheduler, etc.)

Note that those are all use cases for dynamic binding, not thread-local state. If we want to support dynamic binding efficiently then we should just do that directly. Using TLS for it will always be less efficient.

> Why not stick to the same get/set interface as all the other languages, especially if it's also the fast option?

For the same reason OCaml doesn't let you mutate `let` bindings even though other languages do. It improves the ability to reason about the code and makes clear where the actual needed mutation is.

I'm probably in favour of adding a `let mutable` construct, and similarly could be persuaded that there should be a mutable version of thread-local state as well, but I think that if you want to argue for the additional unnecessary mutation in an interface for performance reasons then you do actually need to produce some evidence that it helps with the performance of real programs in practice. Especially when the implementation of `get` isn't particularly optimized right now.

**Reply by c-cube** [reply]
> Note that those are all use cases for dynamic binding, not thread-local state. 

Sure, in absolute terms I'd take a dynamic binding construct as well but what are the odds?? As it stands, TLS is the only reasonable way to implement it in OCaml that I know of, and it's a well known, proven construct for that. 

> I think that if you want to argue for the additional unnecessary mutation in an interface for performance reasons then you do actually need to produce some evidence that it helps with the performance of real programs in practice.

TLS is naturally mutable? I don't see any actual benefit to removing the mutation besides general vibes. It's pretty clear that fewer indirections = better perf in general. 

In my code I have `foobar option ref TLS.key` in various places (because the API is immutable, sth @gadmm has been kind enough to change in a merged but yet to be released PR); I suspect it'd be a common occurrence. The `option` is there to differentiate between initialized and uninitialized data (e.g. if there's need to allocate a local buffer per thread, but lazily); that means that `ref` is actual overhead.

**Reply by lthls** [reply]
If you use mutable TLS, every access is an expensive `TLS.get`. If you use immutable TLS storing a ref, you have one access which is an expensive `TLS.get` when you first enter from an unknown context, and all the remaining accesses are cheap mutable loads until you leave this context.
You might need to tweak your code to take advantage of that, unless your code is in the right shape to benefit from CSE of the immutable `TLS.get` operation.

**Reply by lpw25** [reply]
> In my code I have `foobar option ref TLS.key` in various places

Then you could use [moption](https://ocaml.janestreet.com/ocaml-core/v0.12/doc/core_kernel/Moption/index.html) and have the same number of indirections as if you used `foobar option TLS.key` with a mutable key.

**Reply by kayceesrk** [reply]
I do like the idea of just having a `get`. When programming with DLS, one needs a couple of fields of thread-local state, which are accessed together. For this, using a record with some mutable fields is preferable to accessing each field with a separate key. You only pay to cost of an expensive `TLS.get` once.

**Reply by gadmm** [reply]
> If we want to support dynamic binding efficiently then we should just do that directly. Using TLS for it will always be less efficient.

Could you please expand on how implementing dynamic binding could be more efficient?

**Comment by c-cube** [comment]
- this assumes a compiler optimization, I assume (or that the primitive
  is at least annotated as idempotent)?
- if you have to tweak you code, you can also tweak it in the mutable
  case so it only does one `get`? Maybe an example would help.

**Reply by lthls** [reply]
- Yes, although it is an optimisation that is already done for other kinds of immutable loads.
- Here is an example which shows what I have in mind:
```ocaml
(* immut + ref version *)
let init_val = ...
let key = TLS.make (fun () -> ref init_val)
let get_my_tls () = !(TLS.get key)
let set_my_tls v = TLS.get key := v

let foo () =
  let first_val = get_my_tls () in
  maybe_call set_my_tls;
  let second_val = get_my_tls () in
  ...
```
The two calls to `get_my_tls` (assuming they are inlined) are in the same context, so the second call to `TLS.get key` can reuse the result of the first one (and if `maybe_call` and `set_my_tls` are inlined too, the corresponding call can be shared too).
Tweaking the code would be useful if `get_my_tls` is *not* inlined. Here is what it may look like:
```ocaml
let get_my_tls ?tls_get_key () =
  match tls_get_key with
  | None -> !(TLS.get key)
  | Some r -> !r

let set_my_tls ?tls_get_key v =
  match tls_get_key with
  | None -> TLS.get key := v
  | Some r -> r := v

let foo () =
  let tls_get_key = TLS.get key in
  let first_val = get_my_tls ~tls_get_key () in
  maybe_call (set_my_tls ~tls_get_key);
  let second_val = get_my_tls ~tls_get_key () in
  ...
```

Of course, all of this assumes that `TLS.get` would be a single primitive. I haven't looked at the proposed implementations of `TLS`, but at least for `DLS.get` this is not the case (it is a big ocaml function, the only primitives involved are the ones to get and set the whole DLS storage area for the current domain).

**Comment by c-cube** [comment]
So I need to pull Core as a dependency to get basic TLS functionality? I'm not very enthusiastic about that, honestly.

**Comment by gasche** [comment]
I had a look at existing use-cases for `DLS.set` in OCaml codebases. Here is what I found.

- There are use-cases of settable DLS keys that are arguably performance-sensitive, for example the use in the `Str` module to store the global regexp engine state (which is `set` on each match request, and `get` when the user queries the result), but I think that they would not be sensitive to an indirection through a reference. (My intuition is that indirections are an issue when there are a lot of them to traverse in a given workload, for example an array of references, but less of an issue when there is a small, constant number of them, which just slightly increases cache pressure.)

- I found other use-cases where `DLS.set` is used for configuration. This is notably the case for Domain_local_timeout and Domain_local_await from @polytypic (see https://github.com/ocaml-multicore/domain-local-await/blob/2028ac338517b91fa93072486d03769d56b3f4df/src/Domain_local_await.ml#L96 for example), where the user can call `DLS.set` once per domain to move from per-domain state (which is the default) to per-thread state. It is also used in `ebsl`, an (apparently abandoned) multicore scheduler for OCaml, where DLS.set is used to setup the per-domain state once when initializing a domain, see https://github.com/bartoszmodelski/ebsl/blob/8669675c2824e76cc27be6703bea770ef4975da7/schedulr/scheduler.ml#L181 .

I think that the use-case for configuration suggests a potential issue if we remove `DLS.set` from the API. Currently, the initializer provided by `make` defines a central policy to decide the DLS value on each domain, that is decided globally at key-creation time. When someone spawns a new domain, if they first call `DLS.get` they will get a value computed by this central policy. But they can also call `DLS.set` first to decide on the value themselves, in which case the central initializer is never called on this domain. (It is of course possible to call `DLS.set` several times, but those workflows do not rely on this.)

One potential compromise in the API design space would be to ask that DLS keys be single-assignment: `DLS.set` must be called at most once, before the first call to `DLS.get`. In my API documentation sketch I expose that a key on a given domain can contain a value or be empty; `set` would only be allowed on empty keys. This would validate common-subexpression-elimination of `DLS.get`, without taking away the ability of a domain to initialize its own state independently of the central policy.

There may be other approaches to this local-initialization or decentralized-initialization need:
- It's possible to encode fully explicitly with the current API, by using `foo option ref DLS.t` as @c-cube suggests, or a specialized datatype
- Maybe we could think of a different API that clearly separates the centrally-initialized and the locally-initialized keys

**Comment by gadmm** [comment]
**async-safety**

Whichever the choice regarding `split_from_parent` and the default initialization function, please ensure that TLS and DLS can be read from async callbacks. This does not necessarily mean that `get` should itself be async-safe, but if it is not then please provide alternative reading functions, see e.g. `get_opt` and `get_default` here (where `get_default` can be used to raise a catchable exception): https://github.com/c-cube/thread-local-storage/blob/main/src/thread_local_storage.mli. (The latter API is designed to be without overhead in the fast case.)

**Comment by c-cube** [comment]
I'm releasing `thread-local-storage` 0.2 with contributions from @gadmm, and in the end it's possible to have `get` and `set` with 0 additional indirections or allocations. Refactoring my code and libraries to use it, it turns out that almost all of my use cases are in the form of:

```ocaml
let access_local_foo () =
  match TLS.get_exn k_foo with
  | foo -> foo
  | exception TLS.Not_set ->
    let foo = create_foo() in
    TLS.set k_foo foo; foo
```

So, really, not much opportunity for CSE, and no uneeded indirections

**Reply by gasche** [reply]
I suppose that the opportunities for CSE appear when you perform two computations that call `access_local_foo ()` next to each other, if both calls are inlined. For example, in the stdlib the calls to `Format.*printf` functions will access the format-buffer DLS value, and it is plausible that several calls would be next to each other; they could share the results of the first `DLS.get` call.

(We could get the same optimizations with `TLS.set` as well, if we teach the optimizers that `TLS.{get,set}` behaves like a global reference.)

**Reply by gadmm** [reply]
The use of `get_exn` rather than `get_default` is to avoid the (supposed) overhead of allocating a closure in advance (in general) right?

**Reply by gasche** [reply]
Note: the code example of @c-cube above appears to be equivalent to `TLS.get_default ~default:create_foo k_foo`, which does not allocate either.

**Comment by c-cube** [comment]
Yep, I want as little overhead as I can in these situations. I sometimes have local buffers in the TLS to reduce allocations further.

**Comment by gasche** [comment]
My understanding is that DLS/TLS slots are best thought of as single-assignment references. There are several ways to expose single-assignment references in an API. What the current `get` function does is to hide the assignment part by performing it implicitly on the first `get` -- similarly to lazy thunks. The API of @c-cube and @gadmm exposes to users  that a slot is not yet filled, and users can call `set` in that case. Both APIs do not enforce single-assignment, they offer a `set` function that can be called several times, but I have not found strong use-cases for this in the wild.

Two questions come to mind:

1. What's a good API to expose single-assignment slots? Can we be flexible in where the initialization logic is provided, that is allow to provide it either at key-creation time (as our current API) and at key-access time (as in the thread-local-storage API)?

2. Could the compiler actually take advantage of the single-assignment (SA) nature of these slots for optimizations? SA allows to share the result of two successful calls to `get`, even in presence of unknown/foreign code in between (which is invalid for general references, as a call to `set` could be hidden); I think this is the optimization that @lthls is hoping for, but is it realistic to expect the compiler to learn and use it?

**Reply by gasche** [reply]
One API for single-assignment could be as follows:

```ocaml
type ('init, 'v) saslot
val create : 'init -> ('init, 'v) saslot
val get : default:('init -> 'v) -> ('init, 'v) saslot -> 'v
```

If you instantiate `'init` with `unit -> 'v`, then you get the current API with a creation-time function that decides everything. If you instantiate it with `unit`, then you get the thread-local-storage API where the caller decides everything. Note that even with `unit -> 'v`, the getter can still override the initialization choice, which captures some use-cases of `DLS.set` to override a default function in the standard library.

**Reply by c-cube** [reply]
> Both APIs do not enforce single-assignment, they offer a set function that can be called several times, but I have not found strong use-cases for this in the wild.

There you go:
- [case 1 in moonpool](https://github.com/c-cube/moonpool/blob/simon/move-to-picos/src/core/worker_loop_.ml#L108)
- [case 2 in moonpool](https://github.com/c-cube/moonpool/blob/simon/move-to-picos/src/core/worker_loop_.ml#L119)

here I keep the current fiber and runner in TLS to make it faster to spawn other fibers and access FLS (fiber local storage). It's updated at the beginning and end of every single task running on the thread pool.

**Comment by fpottier** [comment]
Hi! Is there a consensus to deprecate DLS in favor of TLS? (Some messages in this thread seem to suggest this.)

Assuming that DLS is staying, I have a question about it. Its API offers just `get` and `set`. It seems unclear to me how one is supposed to atomically read and update a DLS slot. If one naively uses a sequence "`get`; compute; `set`" then this update is not atomic; there is a danger of being interrupted by another thread that runs on the same domain. Should one take a lock? If so, 1- this seems to go against the spirit of domain-local storage, whose name suggests that locking should not be necessary (but maybe it is just a misconception on my part); and 2- using a global lock, shared by all domains, would be counter-productive, so one would then presumably have to use a domain-local lock; that would result in an inefficient sequence of the form "access DLS to find address of lock; acquire lock; access DLS to get, compute, set; release lock".

**Reply by gadmm** [reply]
We use DLS in a library on the C side (with a hand-made implementation using caching in implicit TLS in C). Our experience is that DLS is useful for sharing resources between threads of a domain that would be too expensive to allocate per-thread, and too inefficient to share globally. This is in fact the initial point for the runtime of having domains as a distinct concept from threads: threads in the same domain share a minor heap, etc. But as you note, in order to give this power to the user, the problem of efficient intra-domain synchronisation must be solved.

For us it works because in C we control where context switches happen. In theory one could provide such control to the OCaml programmer (e.g. by allowing to turn off context switches in an expression). For the moment the only way is to reason about polling locations (e.g. with `[@poll error]`), which is limited.

Efficient TLS seems more important and urgent (and simpler to use) than efficient DLS.
